src/Page.js

```diff
@@ -788,11 +821,13 @@ Page.prototype.generate = function (builtFiles) {
     userDefinedVariablesMap: this.userDefinedVariablesMap,
   };
   return new Promise((resolve, reject) => {
-    markbinder.includeFile(this.sourcePath, fileConfig)
+    fs.readFileAsync(this.sourcePath)
       .then((result) => {
         this.collectFrontMatter(result);
         return this.removeFrontMatter(result);
```

Comment:
⭐⭐

The rearranging of the processing order may break some use cases. For example, in `docs/userGuide/markBindSyntaxOverview.md`, the frontmatter actually contains some nunjucks syntax:

```html
<frontmatter>
  title: "User Guide: {{ title }} "
  layout: userGuide
</frontmatter>
```

This caused a problem with the page title being rendered incorrectly.

`includeFile()` also calls `preprocess()`, but in the new case it seems to be not called before this line? Because this does contain precisely the nunjucks rendering that allows the above to work.

Is it possible to obtain the layout name without having to re-order the sequence?

`docs/userGuide/index.md` is pretty broken too. Not sure what is going on there, the frontmatter seems to be totally left behind.

Reply:
Thanks so much for the review :smile:. I added a fix on this. Will address the rest once we confirmed the approach for this.

The new sequence is this:

(1) Markbind preprocessing
(2) Frontmatter
(3) Expressive Layouts

Within expressive layouts, I instantiated a markbinder and parsed the additional wrapper content within layouts.

Reply:
> The new sequence is this:
> 
> (1) Markbind preprocessing
> (2) Frontmatter
> (3) Expressive Layouts
> 
> Within expressive layouts, I instantiated a markbinder and parsed the additional wrapper content within layouts.

Seems fine for this sequence at the moment, can't think of major issues for this.